pyscriptjs/src/remote_interpreter.ts

```diff
@@ -1,45 +1,55 @@
-import { Interpreter } from './interpreter';
+import type { AppConfig } from './pyconfig';
 import { getLogger } from './logger';
+import { Stdio } from './stdio';
 import { InstallError, ErrorCode } from './exceptions';
-import type { loadPyodide as loadPyodideDeclaration, PyodideInterface, PyProxy } from 'pyodide';
 import { robustFetch } from './fetch';
-import type { AppConfig } from './pyconfig';
-import type { Stdio } from './stdio';
+import type { loadPyodide as loadPyodideDeclaration, PyodideInterface, PyProxy } from 'pyodide';
 
 declare const loadPyodide: typeof loadPyodideDeclaration;
-
 const logger = getLogger('pyscript/pyodide');
 
+export type InterpreterInterface = PyodideInterface | null;
+
 interface Micropip extends PyProxy {
     install: (packageName: string | string[]) => Promise<void>;
     destroy: () => void;
 }
 
-export class PyodideInterpreter extends Interpreter {
+/*
+RemoteInterpreter class is responsible to process requests from the
+`InterpreterClient` class -- these can be requests for installation of
+a package, executing code, etc.
+
+Currently, the only interpreter available is Pyodide as indicated by the
+`InterpreterInterface` type above. This serves as a Union of types of
+different interpreters which will be added in near future.
+
+Methods available handle loading of the interpreter, initialization,
+running code, loading and installation of packages, loading from files etc.
+
+The class will be turned `abstract` in future, to support more runtimes
+such as MicroPython.
+ */
+export class RemoteInterpreter extends Object {
     src: string;
-    stdio: Stdio;
-    name?: string;
-    lang?: string;
-    interface: PyodideInterface;
+    interface: InterpreterInterface;
     globals: PyProxy;
     // TODO: Remove this once `runtimes` is removed!
-    interpreter: PyodideInterface;
+    interpreter: InterpreterInterface;
```

Comment:
this distinction between `RemoteInterpreter` and `InterpreterInterface` seems very over-engineered.
Unless there is a good reason to keep it, I'd like to merge all its functionalities directly into this class.
Once we add support for MicroPython we will likely end up with a `RemoteInterpreter` base class and a subclass for each concrete interpreter, but I cannot imagine why this `interface` would be useful.

Reply:
Can we tackle this in another PR? Right now, the idea is that a `RemoteInterpreter` has an underlying interface which can be `Pyodide` or `MicroPython` or whatever.

And thus, the type of it is a Union of `PyodideInterface` and `null` currently, with `MicroPython` to come in future.

Reply:
I am fine to do this in another PR, as long as we actually do this and don't postpone it forever.

> the idea is that a RemoteInterpreter has an underlying interface which can be Pyodide or MicroPython or whatever.

yes I understand the idea, but I don't really see why we need yet another level of indirection: we can have a `PyodideRemoteInterpreter` and a `MicroPythonRemoteInterpreter` and that's it.

Obligatory quote :)
![image](https://user-images.githubusercontent.com/344507/222779218-0f729da6-8da9-4943-946d-e7ce1402f6d5.png)

Reply:
Haha, sure, thanks @antocuni